src/plasma/plasma_manager.cc

```diff
@@ -642,19 +647,16 @@ int read_object_chunk(ClientConnection *conn, PlasmaRequestBuffer *buf) {
   }
   r = read(conn->fd, buf->data + conn->cursor, s);
 
-  if (r == -1) {
-    LOG_ERROR("read error");
-  } else if (r == 0) {
-    LOG_DEBUG("end of file");
+  if (r <= 0) {
+    LOG_ERROR("Read error");
+    return errno;
   } else {
     conn->cursor += r;
-  }
-  /* If the cursor is equal to the full object size, reset the cursor and we're
-   * done. */
-  if (conn->cursor == buf->data_size + buf->metadata_size) {
-    conn->cursor = 0;
-    return 1;
-  } else {
+    /* If the cursor is equal to the full object size, reset the cursor and
+     * we're done. */
+    if (conn->cursor == buf->data_size + buf->metadata_size) {
+      ClientConnection_finish_request(conn);
+    }
```

Comment:
Should be impossible since we pass `s` into the `read` call above, but might be worth having a sanity check 

```
CHECK(conn->cursor <= buf->data_size + buf->metadata_size);
```